### ISSUE TITLE
Bug with random seed in paccman generator

### ISSUE THREAD

**Issue by itaijj** [GT4SD/gt4sd-core#224]
**Describe the bug**
PaccmannGP does not have deterministic generator behaviour although there is a code that sets a seed here
https://github.com/GT4SD/gt4sd-core/blob/ea9c299cea6ea6c63696b4e9ee49ceb16fb58507/src/gt4sd/algorithms/controlled_sampling/paccmann_gp/core.py#L191
it seems that this seed is only used for the optimizer but not for other sources of randomness in the code
https://github.com/GT4SD/gt4sd-core/blob/ea9c299cea6ea6c63696b4e9ee49ceb16fb58507/src/gt4sd/algorithms/controlled_sampling/paccmann_gp/implementation.py#L237C36-L237C36

I also tried the following after some googling without success,
it seems the seed should also be transferred to some network layer operation which I can't change when the model is fixed.
import numpy as np
import random
import tensorflow as tf

def seed_everything(seed=1234):
    random.seed(seed)
    os.environ['PYTHONHASHSEED'] = str(seed)
    np.random.seed(seed)
    tf.random.set_seed(seed)

def set_global_determinism(seed=1234):
    seed_everything(seed=seed)
    os.environ['TF_DETERMINISTIC_OPS'] = '1'
    os.environ['TF_CUDNN_DETERMINISTIC'] = '1'
    tf.config.threading.set_inter_op_parallelism_threads(1)
    tf.config.threading.set_intra_op_parallelism_threads(1)

**Comment by drugilsberg** [comment]
Thanks for reporting this and analyzing the issue. If we find a solution we will post it here.
In case you find it, feel free to contribute it back following this: https://github.com/GT4SD/gt4sd-core/blob/main/CONTRIBUTING.md

**Comment by jannisborn** [comment]
Hi @itaijj
Thx for the interest in GT4SD. Have you tried fixing all torch random seeds? PaccMannGP uses torch, not TF.

**Comment by itaijj** [comment]
I tried also torch random seeds before that and it didn't work.
I see that you also use gaussian process modules from scikit, 
do you know if it can become deterministic somehow too? 
https://scikit-optimize.github.io/stable/modules/generated/skopt.gp_minimize.html

**Comment by jannisborn** [comment]
`skopt.gp_minimize` has a `random_state` input so, yes!

Maybe try fixing this parameter on top of your current changes. Let us know whether it works

**Comment by itaijj** [comment]
if it's indeed a pytoch model, where is the code that changes it to eval mode?
This is one of the required steps mentioned in web, but I don't find such a code and the model type is not a torch model.

**Comment by jannisborn** [comment]
The sklearn/skopt handling of the random state seems sane. The seed is passed down to the initializer in the code you linked.

A separate source of stochasticity is in the SMILES generator which uses `SamplingSearch` from `paccmann_chemistry` during decoding, see [here](https://github.com/GT4SD/gt4sd-core/blob/ea9c299cea6ea6c63696b4e9ee49ceb16fb58507/src/gt4sd/algorithms/controlled_sampling/paccmann_gp/implementation.py#L128). The source code in paccmann_chemistry is [here](https://github.com/PaccMann/paccmann_chemistry/blob/master/paccmann_chemistry/utils/search.py): 
But this should be affected by setting the global torch seed. As a next step I recommend to verify whether the remaining randomness comes from the skopt component or from the model (smiles generation).
Remember that skopt is called *after* the smiles generator in every optimization step, so if the smiles generator is still stochastic, fixing the seed in skopt wont have any effect.

There's also random jitter being added to the latent point [here](https://github.com/GT4SD/gt4sd-core/blob/ea9c299cea6ea6c63696b4e9ee49ceb16fb58507/src/gt4sd/algorithms/controlled_sampling/paccmann_gp/implementation.py#L261).

**Comment by jannisborn** [comment]
Model is being set in `eval` mode [here](https://github.com/GT4SD/gt4sd-core/blob/main/src/gt4sd/algorithms/controlled_sampling/paccmann_gp/implementation.py#L125)

**Comment by itaijj** [comment]
I see, thanks.
I tried currently to set the global seeds like this before creating and generating using the PaccMannGP model
def seed_everything(seed=42):
    random.seed(seed)
    os.environ['PYTHONHASHSEED'] = str(seed)
    np.random.seed(seed)
    torch.manual_seed(seed)
    torch.backends.cudnn.deterministic = True
    torch.backends.cudnn.benchmark = False

I still get different results each time, any other idea how to set things?

**Comment by itaijj** [comment]
after debugging the code for a while,
it seems the generated smiles remains always the same ( at least for the first round)
but the code that creates the randomness is core sampling that generate in a infinite loop and is conditioned by time.
because the time it takes is not deterministic I think the output is not deteministic
https://github.com/GT4SD/gt4sd-core/blob/ea9c299cea6ea6c63696b4e9ee49ceb16fb58507/src/gt4sd/algorithms/core.py#L225

**Comment by itaijj** [comment]
I found out if ordered smiles_set  has the same molecules,
because we sample num_items from the unordered set we get randomness!

**Comment by itaijj** [comment]
it fixes it, but I don't want to change the code

**Comment by itaijj** [comment]
I can bypass it by having same num_samples and maximum_number_of_sampling_steps

**Comment by jannisborn** [comment]
It does induce a bias too, in this case the molecules are sorted in reverse order by length, so short molecules are ignored.

But note that this is a legacy code snippet that is not executed when running PaccMannRL. Good catch, actually, we can remove this function. The function that is executed is this: https://github.com/GT4SD/gt4sd-core/blob/ea9c299cea6ea6c63696b4e9ee49ceb16fb58507/src/gt4sd/algorithms/conditional_generation/paccmann_rl/implementation.py#L70
which is called from `generate_batch(`

**Comment by jannisborn** [comment]
Submitted a PR in #226 to fix this @itaijj
